docs/spec_draft.md

```diff
@@ -627,6 +628,47 @@ matrix, then the behavior is undefined.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.clamp
+
+### Semantics
+
+Clamps every element of the `operand` tensor between a minimum and maximum
+value and produces a `result` tensor. More formally, `result[i0, ..., iR-1]` =
+`minimum(maximum(operand[i0, ..., iR-1], min[i0, ..., iR-1]), max[i0, ..., iR-1])`,
+where `minimum` and `maximum` operations correspond to
+[stablehlo.minimum](#stablehlominimum) and
+[stablehlo.maximum](#stablehlomaximum).
```

Comment:
Do we need `clamp` in the StableHLO opset (as compared with e.g. CHLO) if it's just a composition of `minimum` and `maximum`? Or there are some floating-point shenanigans involved, and it isn't?

Reply:
Were you referring to issues like mixed precision when talking about floating-point shenanigans?

Reply:
Sorry - I should've been more precise. I meant something different from mixed precision.

Let's take another example first. One might think that `stablehlo.minimum` can be replaced by a combination of `stablehlo.compare` and `stablehlo.select`. However, as illustrated by [elemental_ir_emitter.cc](https://github.com/tensorflow/tensorflow/blob/49959d2f7d5e497edf90d6cdebc9afa3247234b7/tensorflow/compiler/xla/service/llvm_ir/llvm_util.cc#L108-L120), it's not that easy:
  * First, because of the floating-point shenanigans that I was referring to, we need not two but four ops to represent minimum in full generality because we need to properly account for NaNs.
  * Furthermore, depending on compiler flags, sometimes we don't want to properly account for NaNs and want two ops instead.

As a result, I don't think it would be a good idea to not have `minimum` (or `maximum`) ops in StableHLO, because without these dedicated ops, we won't be able to reliably support a wide range of producers (there are producers who care about fastmath, and there are those who don't).

Now, what about `stablehlo.clamp`? From what I can see, [elemental_ir_emitter.cc](https://github.com/tensorflow/tensorflow/blob/49959d2f7d5e497edf90d6cdebc9afa3247234b7/tensorflow/compiler/xla/service/elemental_ir_emitter.cc#L1906-L1927) seems to suggest that it's completely equivalent to a combination of `minimum` and `maximum`. If that's indeed the case, do we need `clamp` as a dedicated op in the StableHLO opset? Or maybe it should be moved to CHLO?

Reply:
1. Minimum/maximum do not support complex datatypes(but that is a different topic of discussing altogether) and implicit broadcasting, but clamp does support implicit broadcasting. 
2. We can still lower it to other ops in stablehlo (broadcast, min and max) and decide to move it to chlo, in this regard what should be our guiding principles - If an op can be completely and correctly lowered to other ops in stablehlo we are free to move it to chlo or are there other nitty-gritties to consider?

Reply:
I think we should have the a clamp operation in StableHLO.

From the [above link](https://github.com/tensorflow/tensorflow/blob/49959d2f7d5e497edf90d6cdebc9afa3247234b7/tensorflow/compiler/xla/service/elemental_ir_emitter.cc#L1906-L1927) , the clamp on integral types seems straightforward, but for the floating point type,  the lowering is based on  `EmitFloatMin` and `EmitFloatMax` which poses the same problem, as you mentioned above, preventing us from getting rid of of max/min.

```
 if (primitive_util::IsFloatingPointType(prim_type)) {
    return EmitFloatMin(max_value, EmitFloatMax(min_value, arg_value, ""), "");
```

Reply:
1) Good catch! Given that we are considering removing support for complex tensors from minimum/maximum, we should remove it from clamp as well using the same rationale. Let's update #560 accordingly? cc @sdasgup3 to confirm.

2) "in this regard what should be our guiding principles - If an op can be completely and correctly lowered to other ops in stablehlo we are free to move it to chlo". Mostly, yes, except that I would replace "are free to" with "it's a good idea", because we need an RFC for changes like that.

3) @sdasgup3 I'm not sure I follow. Sounds like your example actually supports the theory that `clamp` is trivially decomposable to `minimum` and `maximum`?

Reply:
1. Will get back asap.

3. Ah! I got this wrong.  Yes, Stablehlo Max/Min is already doing all the heavy lifting in dealing with all the "floating-point shenanigans". With that the clamp implementation, based on max/min, seems trivially decomposable. 

From the guidelines you mentioned above  it seems like an ideal candidate for that.

Reply:
relevant discussion https://github.com/openxla/stablehlo/pull/495#discussion_r1028776747 on how should be approach the transition of an op to CHLO.

Reply:
1) Left a comment on #560 to not forget about complex types in `clamp`.
2-3) Opened #582 to continue the discussion of potentially moving `clamp` to CHLO.